src/python/pants/binaries/binary_tool.py

```diff
@@ -175,6 +205,22 @@ def _make_binary_request(self, version):
       external_url_generator=self.get_external_url_generator(),
       archiver=self._get_archiver())
 
+  def _get_url_to_fetch(self) -> UrlToFetch:
```

Comment:
Rather than calling private methods from `@rules`, IMO: put the logic in rules, and make classes thinner (ie, closer to being dataclasses with public data).

Reply:
Got it!